From a888cfe36909ea4d7e8ae48a2fc2ef6c8eaf01bc Mon Sep 17 00:00:00 2001
From: Lauri Myllari <lauri.myllari@gmail.com>
Date: Mon, 25 Nov 2013 15:37:25 -0800
Subject: [PATCH] drm/i915: Add "Video 16:235 pass-through" RGB range

This setting is for 0-255 level output signal tagged as limited range in
AVI infoframes. This allows for better video output (BtB, WtW and no
banding).

TODO: only HDMI at the moment
---
 drivers/gpu/drm/i915/i915_drv.h    |    1 +
 drivers/gpu/drm/i915/intel_drv.h   |    7 +++++++
 drivers/gpu/drm/i915/intel_hdmi.c  |   18 ++++++++++++++++--
 drivers/gpu/drm/i915/intel_modes.c |    1 +
 4 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index b12d942..03f9b0a 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -2435,6 +2435,7 @@ int vlv_freq_opcode(int ddr_freq, int val);
 #define INTEL_BROADCAST_RGB_AUTO 0
 #define INTEL_BROADCAST_RGB_FULL 1
 #define INTEL_BROADCAST_RGB_LIMITED 2
+#define INTEL_BROADCAST_RGB_VIDEO 3
 
 static inline uint32_t i915_vgacntrl_reg(struct drm_device *dev)
 {
diff --git a/drivers/gpu/drm/i915/intel_drv.h b/drivers/gpu/drm/i915/intel_drv.h
index 9d2624f..dce51e7 100644
--- a/drivers/gpu/drm/i915/intel_drv.h
+++ b/drivers/gpu/drm/i915/intel_drv.h
@@ -239,6 +239,12 @@ struct intel_crtc_config {
 	 */
 	bool limited_color_range;
 
+	/*
+	 * Set limited color range in AVI infoframe without compressing
+	 * the range.
+	 */
+	bool video_color_range;
+
 	/* DP has a bunch of special case unfortunately, so mark the pipe
 	 * accordingly. */
 	bool has_dp_encoder;
@@ -437,6 +443,7 @@ struct intel_hdmi {
 	int ddc_bus;
 	uint32_t color_range;
 	bool color_range_auto;
+	bool color_range_video;
 	bool has_hdmi_sink;
 	bool has_audio;
 	enum hdmi_force_audio force_audio;
diff --git a/drivers/gpu/drm/i915/intel_hdmi.c b/drivers/gpu/drm/i915/intel_hdmi.c
index 51a8336..78447d6 100644
--- a/drivers/gpu/drm/i915/intel_hdmi.c
+++ b/drivers/gpu/drm/i915/intel_hdmi.c
@@ -373,7 +373,8 @@ static void intel_hdmi_set_avi_infoframe(struct drm_encoder *encoder,
 	}
 
 	if (intel_hdmi->rgb_quant_range_selectable) {
-		if (intel_crtc->config.limited_color_range)
+		if (intel_crtc->config.limited_color_range ||
+		    intel_crtc->config.video_color_range)
 			frame.avi.quantization_range =
 				HDMI_QUANTIZATION_RANGE_LIMITED;
 		else
@@ -889,6 +890,9 @@ bool intel_hdmi_compute_config(struct intel_encoder *encoder,
 	if (intel_hdmi->color_range)
 		pipe_config->limited_color_range = true;
 
+	if (intel_hdmi->color_range_video)
+		pipe_config->video_color_range = true;
+
 	if (HAS_PCH_SPLIT(dev) && !HAS_DDI(dev))
 		pipe_config->has_pch_encoder = true;
 
@@ -1041,25 +1045,35 @@ intel_hdmi_set_property(struct drm_connector *connector,
 	if (property == dev_priv->broadcast_rgb_property) {
 		bool old_auto = intel_hdmi->color_range_auto;
 		uint32_t old_range = intel_hdmi->color_range;
+		bool old_range_video = intel_hdmi->color_range_video;
 
 		switch (val) {
 		case INTEL_BROADCAST_RGB_AUTO:
 			intel_hdmi->color_range_auto = true;
+			intel_hdmi->color_range_video = false;
 			break;
 		case INTEL_BROADCAST_RGB_FULL:
 			intel_hdmi->color_range_auto = false;
 			intel_hdmi->color_range = 0;
+			intel_hdmi->color_range_video = false;
 			break;
 		case INTEL_BROADCAST_RGB_LIMITED:
 			intel_hdmi->color_range_auto = false;
 			intel_hdmi->color_range = HDMI_COLOR_RANGE_16_235;
+			intel_hdmi->color_range_video = true;
+			break;
+		case INTEL_BROADCAST_RGB_VIDEO:
+			intel_hdmi->color_range_auto = false;
+			intel_hdmi->color_range = 0;
+			intel_hdmi->color_range_video = true;
 			break;
 		default:
 			return -EINVAL;
 		}
 
 		if (old_auto == intel_hdmi->color_range_auto &&
-		    old_range == intel_hdmi->color_range)
+		    old_range == intel_hdmi->color_range &&
+		    old_range_video == intel_hdmi->color_range_video)
 			return 0;
 
 		goto done;
diff --git a/drivers/gpu/drm/i915/intel_modes.c b/drivers/gpu/drm/i915/intel_modes.c
index 0e860f3..db39cbb 100644
--- a/drivers/gpu/drm/i915/intel_modes.c
+++ b/drivers/gpu/drm/i915/intel_modes.c
@@ -103,6 +103,7 @@ static const struct drm_prop_enum_list broadcast_rgb_names[] = {
 	{ INTEL_BROADCAST_RGB_AUTO, "Automatic" },
 	{ INTEL_BROADCAST_RGB_FULL, "Full" },
 	{ INTEL_BROADCAST_RGB_LIMITED, "Limited 16:235" },
+	{ INTEL_BROADCAST_RGB_VIDEO, "Video 16:235 pass-through" },
 };
 
 void
-- 
1.7.9.5

